### PR TITLE
EOS-22899: F-97D: cleanup pre-factory- csm_agent

### PR DESCRIPTION
**Pull request by udayan-y21**
Signed-off-by: Udayan Yaragattikar <udayan.yaragattikar@seagate.com>

# Backend

# Problem Statement
https://jts.seagate.com/browse/EOS-22899

Create Pre-factory interfacr in Cleanup phase for  csm-setup

## Design
- Pre-factory flag to be added to Cleanup phase
- Pre-factory will take system back to pre-factory stage i,e. Before post-install phase
- For csm_setup, prefactory will 
                   -- cleanup the service file
                   -- Remove service user in Dev mode only.
- Other functionalities of cleanup will run as it is
## Coding

- _Coding conventions are followed and code is consistent. Yes/No?_Y
- _Confirm All CODACY errors are resolved. Yes/No?_Y

## Testing
Add Pre-factory arg to cleanup command
```
[root@ssc-vm-rhev4-0906 731368]# csm_setup cleanup --help
usage: csm_setup cleanup [-h] [--config CONFIG_URL] [--pre-factory]

optional arguments:
  -h, --help           show this help message and exit
  --config CONFIG_URL  Config Store URL e.g <type>://<path>.
  --pre-factory        Perform pre-factory cleanup
[root@ssc-vm-rhev4-0906 731368]#

```
Testing miniprovisioning steps
```
[root@ssc-vm-rhev4-0906 731368]# csm_setup post_install --config json:///opt/seagate/cortx_configs/provisioner_cluster.json && \
csm_setup prepare --config json:///opt/seagate/cortx_configs/provisioner_cluster.json && \
csm_setup config --config json:///opt/seagate/cortx_configs/provisioner_cluster.json && \
csm_setup init --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
:PASS
:PASS
:PASS

[root@ssc-vm-rhev4-0906 731368]# cat /etc/group  | grep cortxub
wheel:x:10:centos,cortxub
haclient:x:189:hacluster,cortxub
cortxub:x:5007:

[root@ssc-vm-rhev4-0906 731368]# cat /etc/systemd/system/csm_agent.service
# CORTX-CSM: CORTX Management web and CLI interface.
# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
# For any questions about this software or licensing,
# please email opensource@seagate.com or cortx-questions@seagate.com.

[Install]
WantedBy=multi-user.target

[Unit]
Description=Cloud Storage Manager
# Wants=rabbitmq-server.service
# After=rabbitmq-server.service syslog.target
# Requires=rabbitmq-server.service

[Service]
Type=forking
PermissionsStartOnly=true
#ExecStartPre=
ExecStartPre=/usr/bin/mkdir -p /var/run/csm
ExecStartPre=/usr/bin/touch /var/run/csm/csm_agent.pid
ExecStartPre=/usr/bin/chown cortxub:cortxub /var/run/csm/csm_agent.pid
ExecStart=/usr/bin/csm_agent
#ExecStopPost=/sbin/fuser -k 28101/tcp
ExecStopPost=/usr/bin/rm -rf /var/run/csm/csm_agent.pid
PIDFile=/var/run/csm/csm_agent.pid
User=cortxub
# Auto restart option is enabled depending on the system installation type.
#< RESTART_OPTION >
TimeoutStartSec=300
[root@ssc-vm-rhev4-0906 731368]#

[root@ssc-vm-rhev4-0906 731368]# pcs resource disable csm-web && pcs resource disable csm-agent

[root@ssc-vm-rhev4-0906 731368]# csm_setup reset --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS

[root@ssc-vm-rhev4-0906 731368]# csm_setup cleanup --config json:///opt/seagate/cortx_configs/provisioner_cluster.json --pre-factory
:PASS
```
Cortxub user removed from system--> Only in DEV mode
```
[root@ssc-vm-rhev4-0906 731368]# cat /etc/group  | grep cortxub
[root@ssc-vm-rhev4-0906 731368]#

```
csm_agent.service file cleanup
```
[root@ssc-vm-rhev4-0906 731368]# cat /etc/systemd/system/csm_agent.service
# CORTX-CSM: CORTX Management web and CLI interface.
# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
# For any questions about this software or licensing,
# please email opensource@seagate.com or cortx-questions@seagate.com.

[Install]
WantedBy=multi-user.target

[Unit]
Description=Cloud Storage Manager
# Wants=rabbitmq-server.service
# After=rabbitmq-server.service syslog.target
# Requires=rabbitmq-server.service

[Service]
Type=forking
PermissionsStartOnly=true
#ExecStartPre=
ExecStartPre=/usr/bin/mkdir -p /var/run/csm
ExecStartPre=/usr/bin/touch /var/run/csm/csm_agent.pid
ExecStartPre=/usr/bin/chown <USER>:<USER> /var/run/csm/csm_agent.pid
ExecStart=/usr/bin/csm_agent
#ExecStopPost=/sbin/fuser -k 28101/tcp
ExecStopPost=/usr/bin/rm -rf /var/run/csm/csm_agent.pid
PIDFile=/var/run/csm/csm_agent.pid
User=<USER>
# Auto restart option is enabled depending on the system installation type.
#< RESTART_OPTION >
TimeoutStartSec=300


```
- _Confirm that Test Cases are added (for both the cases, fix and feature). Yes/No?_
- _Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability. Yes/No?_
- _Confirm Testing was performed with installed RPM. Yes/No?_
- _Confirm all the ut/regression/smoke test has been performed [Y/N]_

## Checklist

- _PR is self-reviewed? Yes/No?_
- _GitHub Issue is updated_
- _Jira is updated_
  -  _Check if the description is clear and explained._ 
    -  _Check Acceptance Criterion is defined._
      -  _All the tests performed should be mentioned before Resolving a JIRA._
        -  _Verification needs to be done before marked as Closed/Verified_
        - _Any interface change Yes/No?_
        - _Change notified to other gatekeepers: Yes/No?_
        - _Code reviews done and addressed: Yes/No?_
        - _Codacy issues reviewed and addressed: Yes/No?_
        - _Deployment test : Done/Not?_
        - _UT/smoke test: Done/Not?_
        - _Jira number added to PR: Yes/No?_
        - _Github merge done using UI: Yes/No?_
        - _Side effects on other features - deployment/update/node-replacement_
        - _Changes to quick-start-guide/community impact_
        - _All dependent component code PR are raised or already merged Yes/No_
        - _All dependent code already merged in landing branch Yes/No_

### CODE REVIEW

File: csm/conf/cleanup.py [hunk 1]
```diff
@@ -38,24 +38,70 @@ async def execute(self, command):
...
+        await self._unsupported_feature_entry_cleanup()
         self.files_directory_cleanup()
         self.web_env_file_cleanup()
         return Response(output=const.CSM_SETUP_PASS, rc=CSM_OPERATION_SUCESSFUL)
 
+    def _replace_csm_service_file(self):
+        '''
+        Service file cleanup
+        '''
+        Log.info(f"Replace service file.")
```
Comment (pawankumarsrivastava) [inline csm/conf/cleanup.py:58]: Replace -> Replacing

File: csm/conf/cleanup.py [hunk 2]
```diff
@@ -38,24 +38,70 @@ async def execute(self, command):
         try:
             Log.info("Loading configuration")
             Conf.load(const.CSM_GLOBAL_INDEX, const.CSM_CONF_URL)
+            Conf.load(const.DATABASE_INDEX, const.DATABASE_CONF_URL)
         except KvError as e:
             Log.error(f"Configuration Loading Failed {e}")
+        if command.options.get("pre-factory"):
+            # Pre-Factory: Cleanup the system and take to
+            #               pre-factory (Postinstall) stage
```
Comment (pawankumarsrivastava) [inline csm/conf/cleanup.py:46]: Indentation for comment

File: csm/conf/cleanup.py [hunk 3]
```diff
@@ -38,24 +38,70 @@ async def execute(self, command):
...
+        Setup._run_cmd(f"cp -f {const.CSM_AGENT_SERVICE_FILE_SRC_PATH} /etc/systemd/system/")
+
+    def _service_user_cleanup(self):
+        '''
+        Remove service user if system deployed in dev mode.
+        '''
+        self._user = Conf.get(const.CSM_GLOBAL_INDEX, f"{const.CSM}>{const.USERNAME}")
+        if Conf.get(const.CSM_GLOBAL_INDEX, const.KEY_DEPLOYMENT_MODE) == const.DEV and \
+                    self._is_user_exist():
+            Log.info(f"Remove Service user: {self._user}")
```
Comment (pawankumarsrivastava) [inline csm/conf/cleanup.py:68]: Remove -> Removing

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
Minor changes

File: csm/conf/cleanup.py (hunk 1 above)
Comment (udayan-y21) [reply csm/conf/cleanup.py]: Modifications done

File: csm/conf/cleanup.py (hunk 2 above)
Comment (udayan-y21) [reply csm/conf/cleanup.py]: Modifications done

File: csm/conf/cleanup.py (hunk 3 above)
Comment (udayan-y21) [reply csm/conf/cleanup.py]: Modifications done

**Review by soniyamoholkar — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**